### ISSUE TITLE
[OData Api Explorer] Issue with displaying of request example for OData action

### ISSUE THREAD

**Issue by Graph2133** [microsoft/aspnet-api-versioning#667]
Hi! I am facing an issue with generation of examples for swagger ui page. It looks like swagger contains wrong examples for OData collection (entity) actions if they have same name.
I am reusing basic example OData + swashbuckle [SwaggerODataSample](https://github.com/microsoft/aspnet-api-versioning/tree/master/samples/aspnetcore/SwaggerODataSample).

In order to reproduce an issue you should add any actions with same names to different OData entity sets. For example add following code to `OrderModelConfiguration.cs`:
```
 if (apiVersion == ApiVersions.V2)
{
    var action = order.Collection.Action("Validate");

     action.Parameter<Guid>("OrderProp1");
     action.Parameter<int>("OrderProp2");
}
```
and the next lines to `PersonModelConfiguration.cs`:
```
if (apiVersion == ApiVersions.V2)
{
   var action = person.Collection.Action("Validate");
   action.Parameter<int>("PersonProp1");
}
```
After opening swagger page and selection of second api version you can see that for 'Orders' we have action with following example:
![image](https://user-images.githubusercontent.com/23173060/94200609-838b7080-fe6f-11ea-887d-c9e5016de685.png)

'People' api will also have Validate action but People.Default.ValidateParameters will contain wrong example value:
![image](https://user-images.githubusercontent.com/23173060/94200846-f8f74100-fe6f-11ea-8cb6-88132815c16a.png)

People.Default.ValidateParameters should contain only 'PersonProp1'.

For OData People and Order controllers I used following action:

```
        /// <summary>
        /// Validates [*].
        /// </summary>
        /// <param name="parameters">The action parameters.</param>
        /// <returns>None</returns>
        [HttpPost]
        [Consumes("application/json")]
        [ProducesResponseType(Status204NoContent)]
        [ProducesResponseType(Status400BadRequest)]
        [ProducesResponseType(Status404NotFound)]
        public IActionResult Validate(ODataActionParameters parameters)
        {
            return NoContent();
        }
```
It looks like OData actions by itself are generated correctly because if you will navigate to '..../api/$metadata?api-version=2.0'
you can will see that we have both validate actions with correct parameters.
```
<Action Name="Validate" IsBound="true">
    <Parameter Name="bindingParameter" Type="Collection(Microsoft.Examples.Models.Order)"/>
    <Parameter Name="OrderProp1" Type="Edm.Guid" Nullable="false"/>
    <Parameter Name="OrderProp2" Type="Edm.Int32" Nullable="false"/>
</Action>
```
and
```
<Action Name="Validate" IsBound="true">
     <Parameter Name="bindingParameter" Type="Collection(Microsoft.Examples.Models.Person)"/>
     <Parameter Name="PersonProp1" Type="Edm.Int32" Nullable="false"/>
</Action>
```
It looks like OData doesn't have any limitations for actions with same name but maybe I am just missing something, could you please help me ?

Renaming of actions solves an issue but it will be great to have consistent naming across multiple endpoint groups

**Comment by commonsensesoftware** [comment]
I'll have to dig into this, but this is probably a bug. I'm guessing that hash computed for the model is somehow coming up the same for both parameter sets. OData doesn't provide any out of the box facilities for this and all the OpenAPI/Swagger generators cannot handle a dictionary of key/value pairs (I'm mean - how can they?). To make this work, API Versioning has generate non-executable type at runtime that can be handled off and be used by Reflection in other libraries. It does this by generating a type that looks like what the EDM says it should. For an action, the _explored_ object isn't **ODataActionParameters**, it's a dynamic type that looks the way the EDM for the action says it should. If this was done for every request, it would be quite slow and could bloat memory out of control.

As you might imagine, this process is already complex enough as it is. There are huge swaths of combinations that developers might create. It's difficult to create a test matrix that would cover them all. Without further analysis, I suspect this is something that falls into the category of two different actions with overlapping names.

I use my own experience to draw some sense of what people _might_ do. I loath _unbound actions_. There's better ways IMHO. This may account for why the scenario was missed. I don't recall testing it, but I reason that 2 _bound actions_ with the same name will work (unconfirmed).

Thanks for reporting this.

**Comment by commonsensesoftware** [comment]
I can confirm that this is definitely a bug. I was able to repro the problem without any further information. There's several related causes.

There's 3 different ways actions (and I believe functions too) can be declared:

1. Container root (these used to be called _Service Actions_ if I remember right)
2. Unbound entity set actions (e.g. a _Service Action_ associated with an entity set)
3. Bound entity set action (e.g. associated with an entity)

The lookup was only properly considering \#1 and \#3; however, \#2 or \#3 may have worked depending on whether the action name is unique. The name in the EDM does not have a unique name across entity sets. Apparently, this should be resolved by supplying the entity type in the set, if there is one. I'll have to do some more testing, but it seems the order of attempts should be \#3 (if there's an entity set), \#2, and then \#1 (least likely).

I'm working a massive set of changes for OData right now. This fix will be included in the changes. This ultimately comes down to trying to match stuff from the EDM up with the explored actions without a request. OData has a large feature set and huge test matrix. I won't be surprised if a few other edge case like this crop up. Hopefully all, if not most, scenarios for this will be resolved after the change.

**Comment by commonsensesoftware** [comment]
The fix for this is now available in [5.0.0-RC.1](https://www.nuget.org/packages/Microsoft.AspNetCore.OData.Versioning/5.0.0-rc.1). The official release will likely occur after 2 weeks of _burn-in_. If you continue to have a problem, feel free to reopen the issue or create a new one. Thanks.
